### PR TITLE
Add support for structured data (isObject and isArray)

### PR DESCRIPTION
**Pull request by nonplus**
Restify-Validation is great for validating route params, headers, queries and form-submissions, and  simple name/value JSON requests.

This PR adds validation support for JSON requests with structured data (i.e. arrays, objects, arrays of objects, etc.) by adding **isArray** (minLength, maxLength, element) and **isObject** (properties) validation configurations:

```javascript
server.put({url: '/products/:id/labels/:label', validation: {
  content: {
    person: {
        isObject: {
            properties: {
                firstName: { isRequired: true },
                middle: { isRequired: false },
                lastName: { isRequired: true },
                emails: {
                    isArray: {
                        maxLength: 5,
                        element: { isEmail: true }
                    }
                }
            }
        }
    },
    label: { isRequired: true }
  }
}}, function (req, res, next) {
    res.send(req.params);
});
```
The README.md contains more details and examples.

### CODE REVIEW

**Comment by gchauvet** [comment]
Gold stuff, thx

File: lib/validation.js [hunk 1]
```diff
@@ -152,48 +166,60 @@ module.exports.process = function (validationModel, req, options) {
     var errors = options.errorsAsArray === false ? {} : [];
     var self = this;
 
-    function addError(error, key) {
```
Comment (nonplus) [inline lib/validation.js]: The `process` function was refactored and most of its work is now done in `validateObject` to allow recursive validation.

File: lib/validation.js [hunk 2]
```diff
@@ -152,48 +166,60 @@ module.exports.process = function (validationModel, req, options) {
...
-                    }
-    	        });
-    	    }
+            self.validateObject("", [], req[realScope], validationRulesScope, validationModel, realScope, req, options, errors);
         }
     });
     return errors;
 };
+
+function addError(errors, error, key, options) {
```
Comment (nonplus) [inline lib/validation.js]: Pulled out because it's now called from `process` and `validateObject`.

File: lib/validators.js [hunk 3]
```diff
@@ -61,6 +61,11 @@ var validators = module.exports = {
                 myValidator.msg = 'Too many elements';
                 return true;
             }
+            if (isArray.element && submittedValue) {
+                return _.reduce(submittedValue, function(previousError, value, index) {
+                    return previousError || this.validateValue('[' + index + ']', _.constant(value), isArray.element);
```
Comment (nonplus) [inline lib/validators.js]: Validate individual elements in the array.

File: lib/validators.js [hunk 4]
```diff
@@ -73,6 +78,10 @@ var validators = module.exports = {
         } else if (!_.isObject(submittedValue) || _.isArray(submittedValue)) {
             myValidator.msg = 'Field is not object';
             return true;
+        } else if (_.isObject(isObject)) {
+            if (isObject.properties) {
+                return this.validateProperties(submittedValue, isObject.properties);
```
Comment (nonplus) [inline lib/validators.js]: Validate the object properties.

**Comment by nonplus** [comment]
@gchauvet, this PR resolves #39 which can be closed.

**Pull request merged**